### PR TITLE
Nodeinfostorage implementation

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: src/database/database.cpp [hunk 1]
```diff
@@ -146,6 +174,95 @@ Database::~Database()
...
+}
+
+aos::Error Database::GetAllNodeIds(aos::Array<aos::StaticString<aos::cNodeIDLen>>& ids) const
+{
+    try {
+        Poco::Data::Statement    statement {*mSession};
+        std::vector<std::string> storedIds;
+
+        statement << "SELECT id FROM nodeinfo;", into(storedIds);
+        if (statement.execute() > ids.MaxSize()) {
```
Comment (mlohvynenko) [inline src/database/database.cpp:236]: Consider checking PushBack result instead of this check

File: src/database/database.cpp [hunk 2]
```diff
@@ -192,3 +315,160 @@ void Database::FromAosCertInfo(const CertInfo& certInfo, aos::iam::certhandler::
...
+        aos::CPUInfo dstItem;
+
+        dstItem.mID         = cpuInfo->getValue<size_t>("id");
+        dstItem.mModelName  = cpuInfo->getValue<std::string>("modelName").c_str();
+        dstItem.mNumCores   = cpuInfo->getValue<size_t>("numCores");
+        dstItem.mNumThreads = cpuInfo->getValue<size_t>("numThreads");
+        dstItem.mArch       = cpuInfo->getValue<std::string>("arch").c_str();
+        dstItem.mArchFamily = cpuInfo->getValue<std::string>("archFamily").c_str();
+
+        dst.PushBack(dstItem);
```
Comment (mlohvynenko) [inline src/database/database.cpp:394]: Consider adding check for PushBack result

File: src/database/database.cpp [hunk 3]
```diff
@@ -192,3 +315,160 @@ void Database::FromAosCertInfo(const CertInfo& certInfo, aos::iam::certhandler::
...
+        AssertNotNull(types);
+        for (const auto& type : *types) {
+            dstItem.mTypes.PushBack(type.convert<std::string>().c_str());
+        }
+
+        dstItem.mTotalSize = partitionInfo->getValue<size_t>("totalSize");
+        dstItem.mPath      = partitionInfo->getValue<std::string>("path").c_str();
+        dstItem.mUsedSize  = partitionInfo->getValue<size_t>("usedSize");
+
+        dst.PushBack(dstItem);
```
Comment (mlohvynenko) [inline src/database/database.cpp:441]: Consider adding check for PushBack result

File: src/database/database.cpp [hunk 4]
```diff
@@ -192,3 +315,160 @@ void Database::FromAosCertInfo(const CertInfo& certInfo, aos::iam::certhandler::
...
+    for (const auto& srcItem : src) {
+        const auto& attribute = srcItem.extract<Poco::JSON::Object::Ptr>();
+        AssertNotNull(attribute);
+
+        aos::NodeAttribute dstItem;
+
+        dstItem.mName  = attribute->getValue<std::string>("name").c_str();
+        dstItem.mValue = attribute->getValue<std::string>("value").c_str();
+
+        dst.PushBack(dstItem);
```
Comment (mlohvynenko) [inline src/database/database.cpp:472]: Consider adding check for PushBack result

File: src/database/nodeinfostorage.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        aos::CPUInfo dstItem;
+
+        dstItem.mID         = cpuInfo->getValue<size_t>("id");
+        dstItem.mModelName  = cpuInfo->getValue<std::string>("modelName").c_str();
+        dstItem.mNumCores   = cpuInfo->getValue<size_t>("numCores");
+        dstItem.mNumThreads = cpuInfo->getValue<size_t>("numThreads");
+        dstItem.mArch       = cpuInfo->getValue<std::string>("arch").c_str();
+        dstItem.mArchFamily = cpuInfo->getValue<std::string>("archFamily").c_str();
+
+        dst.PushBack(dstItem);
```
Comment (mlohvynenko) [inline src/database/nodeinfostorage.cpp:129]: Consider adding check for PushBack result

File: src/database/nodeinfostorage.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        AssertNotNull(types);
+        for (const auto& type : *types) {
+            dstItem.mTypes.PushBack(type.convert<std::string>().c_str());
+        }
+
+        dstItem.mTotalSize = partitionInfo->getValue<size_t>("totalSize");
+        dstItem.mPath      = partitionInfo->getValue<std::string>("path").c_str();
+        dstItem.mUsedSize  = partitionInfo->getValue<size_t>("usedSize");
+
+        dst.PushBack(dstItem);
```
Comment (mlohvynenko) [inline src/database/nodeinfostorage.cpp:176]: Consider adding check for PushBack result

File: src/database/nodeinfostorage.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    for (const auto& srcItem : src) {
+        const auto& attribute = srcItem.extract<Poco::JSON::Object::Ptr>();
+        AssertNotNull(attribute);
+
+        aos::NodeAttribute dstItem;
+
+        dstItem.mName  = attribute->getValue<std::string>("name").c_str();
+        dstItem.mValue = attribute->getValue<std::string>("value").c_str();
+
+        dst.PushBack(dstItem);
```
Comment (mlohvynenko) [inline src/database/nodeinfostorage.cpp:207]: Consider adding check for PushBack result

File: src/iamclient/iamclient.hpp [hunk 8]
```diff
@@ -8,142 +8,128 @@
...
 
     /**
-     * Returns remote node identifiers.
+     * Called when provisioning starts.
      *
-     * @result aos::StaticArray<aos::StaticString<aos::cNodeIDLen>, aos::cMaxNumNodes>
+     * @returns Error.
      */
-    aos::StaticArray<aos::StaticString<aos::cNodeIDLen>, aos::cMaxNumNodes> GetRemoteNodes() override;
+    aos::Error OnStartProvisioning(const aos::String& password) override;
```
Comment (mlohvynenko) [inline src/iamclient/iamclient.hpp:58]: add doxygen comment for param

File: src/database/nodeinfostorage.cpp [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+#include "log.hpp"
+#include "nodeinfostorage.hpp"
+
+using namespace Poco::Data::Keywords;
+
+/***********************************************************************************************************************
+ * Statics
+ **********************************************************************************************************************/
+
+inline std::string Stringify(const Poco::JSON::Object& json)
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/nodeinfostorage.cpp:21]: yes, done

File: src/database/database.cpp [hunk 10]
```diff
@@ -146,6 +174,95 @@ Database::~Database()
...
 
+/***********************************************************************************************************************
+ * nodemanager::NodeInfoStorageItf implementation
+ **********************************************************************************************************************/
+
+aos::Error Database::SetNodeInfo(const aos::NodeInfo& info)
+{
+    try {
+        Poco::JSON::Object pocoNodeInfo;
+
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/database.cpp:185]: done

File: src/database/database.cpp [hunk 11]
```diff
@@ -146,6 +174,95 @@ Database::~Database()
...
+        const auto nodeInfo = Stringify(ConvertNodeInfoToJSON(info));
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/database.cpp:186]: done

File: src/database/database.cpp [hunk 12]
```diff
@@ -7,12 +7,36 @@
...
+{
+    std::ostringstream oss;
+
+    Poco::JSON::Stringifier::stringify(json, oss);
+
+    return oss.str();
+}
+
+template <typename Ptr>
+inline void AssertNotNull(const Ptr& ptr)
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/database.cpp:33]: yes, I changed to return an error

File: src/database/database.cpp [hunk 13]
```diff
@@ -146,6 +174,95 @@ Database::~Database()
...
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error Database::GetNodeInfo(const aos::String& nodeId, aos::NodeInfo& nodeInfo) const
+{
+    try {
+        Poco::Data::Statement statement {*mSession};
+
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/database.cpp:202]: done

File: src/iamclient/iamclient.hpp (hunk 8 above)
Comment (mykola-kobets-epam) [reply src/iamclient/iamclient.hpp:58]: fixed

File: src/database/nodeinfostorage.cpp (hunk 6 above)
Comment (mykola-kobets-epam) [reply src/database/nodeinfostorage.cpp:176]: done

File: src/database/nodeinfostorage.cpp (hunk 5 above)
Comment (mykola-kobets-epam) [reply src/database/nodeinfostorage.cpp:129]: done

File: src/database/database.cpp (hunk 4 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:472]: done

File: src/database/database.cpp (hunk 3 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:441]: done

File: src/database/database.cpp (hunk 2 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:394]: done

File: src/database/database.cpp (hunk 1 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:236]: done

File: src/database/nodeinfostorage.cpp (hunk 7 above)
Comment (mykola-kobets-epam) [reply src/database/nodeinfostorage.cpp:207]: done

**Review by mlohvynenko — APPROVED** [review]

File: src/database/database.cpp [hunk 14]
```diff
@@ -146,6 +166,101 @@ Database::~Database()
...
+            }
+
+            auto err = ConvertNodeInfoFromJSON(*ptr, nodeInfo);
+            if (!err.IsNone()) {
+                return err;
+            }
+        }
+
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << "Failed to get node info: " << e.what();
```
Comment (al1img) [inline src/database/database.cpp:218]: ditto

File: src/database/database.cpp [hunk 15]
```diff
@@ -146,6 +166,101 @@ Database::~Database()
...
+        for (const auto& id : storedIds) {
+            auto err = ids.PushBack(id.c_str());
+            if (!err.IsNone()) {
+                return err;
+            }
+        }
+
+        return aos::ErrorEnum::eNone;
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << "Failed to get node status: " << e.what();
```
Comment (al1img) [inline src/database/database.cpp:245]: ditto

File: src/database/database.cpp [hunk 16]
```diff
@@ -146,6 +166,101 @@ Database::~Database()
...
+        return aos::ErrorEnum::eFailed;
+    }
+}
+
+aos::Error Database::RemoveNodeInfo(const aos::String& nodeId)
+{
+    try {
+        *mSession << "DELETE FROM nodeinfo WHERE id = ?;", bind(nodeId.CStr()), now;
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << "Failed to remove node info: " << e.what();
```
Comment (al1img) [inline src/database/database.cpp:256]: ditto

**Review by mlohvynenko — APPROVED** [review]

File: src/database/database.cpp [hunk 17]
```diff
@@ -146,6 +166,101 @@ Database::~Database()
...
+aos::Error Database::SetNodeInfo(const aos::NodeInfo& info)
+{
+    try {
+        Poco::JSON::Object pocoNodeInfo;
+        const auto         nodeInfo = Stringify(ConvertNodeInfoToJSON(info));
+
+        *mSession << "INSERT INTO nodeinfo (id, info) VALUES (?, ?);", bind(info.mID.CStr()), bind(nodeInfo), now;
+
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << "Failed to set node info: " << e.what();
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/database.cpp:182]: done

File: src/database/database.cpp (hunk 14 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:218]: done

File: src/database/database.cpp (hunk 15 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:245]: done

File: src/database/database.cpp (hunk 16 above)
Comment (mykola-kobets-epam) [reply src/database/database.cpp:256]: done

File: src/database/database.cpp [hunk 18]
```diff
@@ -146,6 +166,101 @@ Database::~Database()
...
+ * nodemanager::NodeInfoStorageItf implementation
+ **********************************************************************************************************************/
+
+aos::Error Database::SetNodeInfo(const aos::NodeInfo& info)
+{
+    try {
+        Poco::JSON::Object pocoNodeInfo;
+        const auto         nodeInfo = Stringify(ConvertNodeInfoToJSON(info));
+
+        *mSession << "INSERT INTO nodeinfo (id, info) VALUES (?, ?);", bind(info.mID.CStr()), bind(nodeInfo), now;
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply src/database/database.cpp:179]: done

**Review by mlohvynenko — APPROVED** [review]

**Review by al1img — APPROVED** [review]

**Pull request merged**